### PR TITLE
genericcontrolplane: split generic config construction from kube-apiserver and apiextensions

### PR DESCRIPTION
**Pull request by sttts**
- follow options -(fiddle)-> config -(fiddle)-> server pattern with generic apiserver config, i.e. split generation appart from apiserver creation
- restore flags and validation logic
- add CompletedServerRunOptions to make completion composable for consumers in their option structs

### CODE REVIEW

File: pkg/genericcontrolplane/options/flags.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+package options
+
+import (
+	cliflag "k8s.io/component-base/cli/flag"
+)
+
+// Flags returns flags for a specific APIServer by section name
+func (s *ServerRunOptions) Flags() (fss cliflag.NamedFlagSets) {
+	// TODO: move all these back into genericcontrolplane
```
Comment (ncdc) [inline pkg/genericcontrolplane/options/flags.go]:
```suggestion
```

File: pkg/genericcontrolplane/server.go [hunk 2]
```diff
@@ -55,6 +56,7 @@ import (
 	"k8s.io/kubernetes/pkg/kubeapiserver"
 	"k8s.io/kubernetes/pkg/serviceaccount"
 	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+	// for workqueue metric registration
```
Comment (ncdc) [inline pkg/genericcontrolplane/server.go]: openshift-goimports fun, I presume?

File: pkg/genericcontrolplane/server.go [hunk 3]
```diff
@@ -79,9 +92,9 @@ func Run(completeOptions completedServerRunOptions, stopCh <-chan struct{}) erro
 	apiExtensionsConfig, err := CreateAPIExtensionsConfig(
 		*config.GenericConfig,
 		config.ExtraConfig.VersionedInformers,
-		pluginInitializer,
-		completeOptions.ServerRunOptions,
-		serviceResolver,
+		nil,
```
Comment (ncdc) [inline pkg/genericcontrolplane/server.go]:
```suggestion
		nil, // pluginInitializer
```

File: pkg/genericcontrolplane/server.go [hunk 4]
```diff
@@ -192,20 +197,16 @@ func CreateKubeAPIServerConfig(s completedServerRunOptions) (
...
+		additionalPluginInitializers...); err != nil {
+		return nil, err
 	}
 
-	// if err := config.GenericConfig.AddPostStartHook("start-kube-apiserver-admission-initializer", admissionPostStartHook); err != nil {
-	// 	return nil, nil, nil, err
-	// }
-
-	// // Load the public keys.
+	// // Load the public keys.g
```
Comment (ncdc) [inline pkg/genericcontrolplane/server.go]:
```suggestion
	// // Load the public keys.
```

File: pkg/genericcontrolplane/server.go [hunk 5]
```diff
@@ -312,16 +308,20 @@ func BuildGenericConfig(
...
 	}
 
-	genericConfig.Authorization.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(s, versionedInformers)
+	genericConfig.Authorization.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(&o.ServerRunOptions, versionedInformers)
 	if err != nil {
 		lastErr = fmt.Errorf("invalid authorization config: %v", err)
 		return
 	}
 
+	//if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIPriorityAndFairness) && s.GenericServerRunOptions.EnablePriorityAndFairness {
```
Comment (ncdc) [inline pkg/genericcontrolplane/server.go]: Why is this added and commented out? A placeholder so we don't lose track of it?

File: pkg/genericcontrolplane/server.go (hunk 2 above)
Comment (ncdc) [reply pkg/genericcontrolplane/server.go]: @sttts bump 😄

File: pkg/genericcontrolplane/server.go (hunk 3 above)
Comment (sttts) [reply pkg/genericcontrolplane/server.go]: Bildschirmfoto 2022-02-08 um 23 37 52

Goland is clever.

Fixed.

File: pkg/genericcontrolplane/server.go (hunk 5 above)
Comment (sttts) [reply pkg/genericcontrolplane/server.go]: I had it in kcp first (because these had not been copied from kube-apiserver here). There we didn't want all the kube, mostly kubelet focused feature gates. Now with the filtering we can removes these comments.

Comment (sttts) [reply pkg/genericcontrolplane/server.go]: done

**Comment by ncdc** [comment]
@sttts we can close this since it's against 1.22 and #40 replaced it for 1.23?

**Comment by sttts** [comment]
For a moment I was confused why this is still open and kcp compiles :D
